dataclass_io/writer.py

```diff
@@ -144,10 +108,89 @@ def writeall(self, dataclass_instances: Iterable[DataclassInstance]) -> None:
 
+    @classmethod
+    @contextmanager
+    def open(
+        cls,
+        filename: str | Path,
+        dataclass_type: type[DataclassInstance],
+        mode: str = "write",
+        overwrite: bool = True,
+        delimiter: str = "\t",
+        comment_prefix: str = "#",
+        **kwds: Any,
+    ) -> Iterator["DataclassWriter"]:
+        """
+        Open a new `DataclassWriter` from a file path.
+
+        Args:
+            filename: The path to the file to which dataclass instances will be written.
+            dataclass_type: The dataclass type to write to file.
+            mode: Either `"write"` or `"append"`.
+                - If `"write"`, the specified file `path` must not already exist unless
+                `overwrite=True` is specified.
+                - If `"append"`, the specified file `path` must already exist and contain a header
+                row matching the specified dataclass and any specified `include_fields` or
+                `exclude_fields`.
+            overwrite: If `True`, and `mode="write"`, the file specified at `path` will be
+                overwritten if it exists.
+            delimiter: The output file delimiter.
+            comment_prefix: The prefix for any comment/preface rows preceding the header row.
+                (This argument is ignored when `mode="write"`. It is used when `mode="append"` to
+                validate that the existing file's header matches the specified dataclass.)
+            **kwds: Additional keyword arguments to be passed to the `DataclassWriter` constructor.
+
+        Yields:
+            A `DataclassWriter` instance.
+
+        Raises:
+            TypeError: If the provided type is not a dataclass.
+            FileNotFoundError: If the output file does not exist when trying to append.
+            IsADirectoryError: If the output file path is a directory.
+            PermissionError: If the output file is not writable.
+            PermissionError: If `mode="append"` and the output file is not readable. (The output
+                file must be readable in order to validate that the existing file's header matches
+                the dataclass's fields.)
+        """
+
+        filepath: Path = Path(filename)
+
+        try:
+            write_mode = WriteMode(mode)
+        except ValueError:
+            raise ValueError(f"`mode` must be either 'write' or 'append': {mode}") from None
+
+        assert_dataclass_is_valid(dataclass_type)
+        if write_mode is WriteMode.WRITE:
+            assert_file_is_writable(filepath, overwrite=overwrite)
+        else:
+            assert_file_is_appendable(filepath, dataclass_type=dataclass_type)
+            assert_file_header_matches_dataclass(
+                file=filepath,
+                dataclass_type=dataclass_type,
+                delimiter=delimiter,
+                comment_prefix=comment_prefix,
+            )
+
+        fout = filepath.open(write_mode.abbreviation)
```

Comment:
This is typed as `IO[Any]`, I assume because we're not passing a literal string as the mode. I tried casting to `TextIOWrapper`, but this yielded `TypeError: write() argument must be str, not bytes`. Couldn't figure out why, but it was late